crates/uv-auth/src/credentials.rs

```diff
@@ -0,0 +1,197 @@
+use base64::prelude::BASE64_STANDARD;
+use base64::write::EncoderWriter;
+use netrc::Authenticator;
+use reqwest::header::HeaderValue;
+use std::io::Write;
+use url::Url;
+
+#[derive(Clone, Debug, PartialEq)]
+pub(crate) struct Credentials {
+    username: String,
+    password: Option<String>,
+}
+
+impl Credentials {
+    pub fn new(username: String, password: Option<String>) -> Self {
+        Self { username, password }
+    }
+
+    pub fn username(&self) -> &str {
+        &self.username
+    }
+
+    pub fn password(&self) -> Option<&str> {
+        self.password.as_deref()
+    }
+
+    /// Extract credentials from a URL.
+    ///
+    /// Returns `None` if `username` and `password` are not populated.
+    pub fn from_url(url: &Url) -> Option<Self> {
+        if url.username().is_empty() && url.password().is_none() {
+            return None;
+        }
+        Some(Self {
+            // Remove percent-encoding from URL credentials
+            // See <https://github.com/pypa/pip/blob/06d21db4ff1ab69665c22a88718a4ea9757ca293/src/pip/_internal/utils/misc.py#L497-L499>
+            username: urlencoding::decode(url.username())
+                .expect("An encoded username should always decode")
+                .into_owned(),
+            password: url.password().map(|password| {
+                urlencoding::decode(password)
+                    .expect("An encoded password should always decode")
+                    .into_owned()
+            }),
+        })
+    }
+}
+
+impl From<Authenticator> for Credentials {
+    fn from(auth: Authenticator) -> Self {
+        Credentials {
+            username: auth.login,
+            password: Some(auth.password),
+        }
+    }
+}
+
+impl Credentials {
+    /// Attach the credentials to the given request.
+    ///
+    /// Any existing credentials will be overridden.
```

Comment:
I thought that request headers was a multi-map, but it looks like the docs say this _will_ override (which we _need_).

Reply:
Ah okay, `append` would append, but `insert` replaces, good.

Reply:
We actually skip if there's an AUTHORIZATION header in the middleware so this isn't critical, but yeah.